### PR TITLE
Improve label vertical align for 2d-tasks/issues/886

### PR DESCRIPTION
**Pull request by knoxHuang**
Re: cocos-creator/2d-tasks#886

image

Changes:
 * 完善 label 垂直对齐问题

### CODE REVIEW

File: cocos2d/core/renderer/utils/label/ttf.js [hunk 1]
```diff
@@ -229,16 +230,20 @@ module.exports = {
...
         else if (_vAlign === macro.VerticalTextAlignment.CENTER) {
-            firstLinelabelY = _canvasSize.height / 2 - lineHeight * (lineCount - 1) / 2;
+            firstLinelabelY = (_canvasSize.height - drawStartY) * 0.5 + _drawFontSize * MIDDLE_RATIO;
         }
         else {
-            firstLinelabelY = _canvasSize.height - lineHeight * (lineCount - 1);
+            firstLinelabelY = _canvasSize.height - drawStartY - _drawFontSize * BASELINE_RATIO - _margin;
         }
 
+        console.log(firstLinelabelY);
```
Comment (pandamicro) [inline cocos2d/core/renderer/utils/label/ttf.js]: 删除调试注释

File: cocos2d/core/renderer/utils/label/ttf.js [hunk 2]
```diff
@@ -176,7 +180,8 @@ module.exports = {
         
         _string = comp.string.toString();
         _fontSize = comp._fontSize;
-        _drawFontsize = _fontSize;
+        _drawFontSize = _fontSize;
+        _underlineWidth = _drawFontSize / 8;
```
Comment (pandamicro) [inline cocos2d/core/renderer/utils/label/ttf.js]: 是 height 吧？？？

File: cocos2d/core/renderer/utils/label/ttf.js [hunk 3]
```diff
@@ -229,16 +230,20 @@ module.exports = {
...
+        let firstLinelabelY = 0;
+        let lineHeight = this._getLineHeight();
+        let drawStartY = lineHeight * (_splitedStrings.length - 1);
         if (_vAlign === macro.VerticalTextAlignment.TOP) {
-            firstLinelabelY = 0;
+            firstLinelabelY = lineHeight + _margin;
         }
         else if (_vAlign === macro.VerticalTextAlignment.CENTER) {
-            firstLinelabelY = _canvasSize.height / 2 - lineHeight * (lineCount - 1) / 2;
+            firstLinelabelY = (_canvasSize.height - drawStartY) * 0.5 + _drawFontSize * MIDDLE_RATIO;
```
Comment (pandamicro) [inline cocos2d/core/renderer/utils/label/ttf.js:243]: 这里貌似还要加上 _margin

File: cocos2d/core/renderer/utils/label/ttf.js [hunk 4]
```diff
@@ -251,70 +256,54 @@ module.exports = {
         //use round for line join to avoid sharp intersect point
         _context.lineJoin = 'round';
         _context.fillStyle = `rgba(${_color.r}, ${_color.g}, ${_color.b}, ${_color.a / 255})`;
-        let underlineStartPosition;
+        let underlineStartPosition = cc.v2();
```
Comment (pandamicro) [inline cocos2d/core/renderer/utils/label/ttf.js]: 能不能不创建临时对象

File: cocos2d/core/renderer/utils/label/ttf.js [hunk 5]
```diff
@@ -251,70 +256,54 @@ module.exports = {
...
 
         //do real rendering
+        let textPos = cc.v2();
```
Comment (pandamicro) [inline cocos2d/core/renderer/utils/label/ttf.js]: 同上

**Review by pandamicro — CHANGES_REQUESTED** [review]

File: cocos2d/core/renderer/utils/label/ttf.js (hunk 3 above)
Comment (knoxHuang) [reply cocos2d/core/renderer/utils/label/ttf.js:243]: 计算 center 不需要加上 _margin

**Review by pandamicro — APPROVED** [review]

File: cocos2d/core/renderer/utils/label/ttf.js [hunk 6]
```diff
@@ -251,70 +258,52 @@ module.exports = {
...
+            _context.fillText(_splitedStrings[i], _drawTextPos.x, _drawTextPos.y);
         }
 
         _texture.handleLoadedTexture();
     },
 
     _calDynamicAtlas (comp) {
-        if(!comp.batchAsBitmap) return;
-        
+        if(!comp.cacheAsBitmap) return;
```
Comment (jareguo) [inline cocos2d/core/renderer/utils/label/ttf.js:293]: 这个错了吧？现在是 batchAsBitmap

**Comment by unace** [comment]
Hello, I tested label vertical alignment with Cocos Creator 2.1.1 (contains this fix) and 2.0.9 (not contain it)
Below is the result.

2.0.9 Editor
image
2.0.9 Web (Mac + Chrome Version 74.0.3729.169)
image
2.0.9 Web (Mac + Safari Version 12.1.1)
image
2.0.9 Android (Pixel2 + Android version 9)
image
2.0.9 Sample project
[label_test_209.zip](https://github.com/cocos-creator/engine/files/3283589/label_test_209.zip)

2.1.1 Editor (Mac)
image
2.1.1 Web (Mac + Chrome Version 74.0.3729.169)
image
2.1.1 Web (Mac + Safari Version 12.1.1)
image
2.1.1 Android (Pixel2 + Android version 9)
image
2.1.1 Sample project
[label_test_211.zip](https://github.com/cocos-creator/engine/files/3283541/label_test_211.zip)

Web Chrome alignment issue has been fixed, but there is a new Android alignment issue.

**Comment by asimaruk** [comment]
Also aligning was changed between 2.1.0 and 2.1.1. Looks like it was good and now broken. Here is an example https://github.com/asimaruk/cc-2.1.1-label-align-issue 1
Reproduces in editor and chrome at least on macos

**Comment by jareguo** [comment]
Thanks guys. We will test your demos to ensure they did fixed in 2.1.2.

**Comment by unace** [comment]
@jareguo Thanks!

**Comment by knoxHuang** [comment]
hi, the bug fix to 2.2  

the engine pr: https://github.com/cocos-creator/engine/pull/4378
the cocos2d-lite pr : https://github.com/cocos-creator/cocos2d-x-lite/pull/1739

**Comment by unace** [comment]
Thanks for the quick fix! Do you have a plan to apply this fix to v2.1.2?

**Comment by jareguo** [comment]
Yes, we will sync to 2.1.2
https://github.com/cocos-creator/engine/pull/4765
https://github.com/cocos-creator/cocos2d-x-lite/pull/1764

Could you possibly have a try?

File: cocos2d/core/renderer/utils/label/ttf.js [hunk 7]
```diff
@@ -353,17 +342,7 @@ module.exports = {
...
-            vAlign = 'top';
-        }
-        else if (_vAlign === macro.VerticalTextAlignment.CENTER) {
-            vAlign = 'middle';
-        }
-        else {
-            vAlign = 'bottom';
-        }
-        _context.textBaseline = vAlign;
+        _context.textBaseline = 'alphabetic';
```
Comment (dumganhar) [inline cocos2d/core/renderer/utils/label/ttf.js:345]: 这个地方为什么写死 textBaseline 为 'alphabetic' ？

File: cocos2d/core/renderer/utils/label/ttf.js [hunk 8]
```diff
@@ -34,14 +34,17 @@ const Overflow = Label.Overflow;
 const WHITE = cc.Color.WHITE;
 const OUTLINE_SUPPORTED = cc.js.isChildClassOf(LabelOutline, Component);
 
+const BASELINE_RATIO = 0.25;
+const MIDDLE_RATIO = (BASELINE_RATIO + 1) / 2 - BASELINE_RATIO;
```
Comment (dumganhar) [inline cocos2d/core/renderer/utils/label/ttf.js:38]: Magic number?

**Pull request merged**